server/src/main/java/io/pravega/schemaregistry/service/SchemaRegistryService.java

```diff
@@ -0,0 +1,924 @@
+                                           Collectors.toMap(AbstractMap.SimpleEntry::getKey, x -> x.getValue().get()));
+                                   return new MapWithToken<>(
+                                           collect, groupsList.getKey());
+                               });
+    }
+
+    /**
+     * Creates new group. Idempotent behaviour. If group already exists, it returns false.
+     *
+     * @param namespace       namespace for which the request is scoped to.
+     * @param group           Name of group.
+     * @param groupProperties Group properties.
+     * @return CompletableFuture which is completed when create group completes. True indicates this was
+     * new group, false indicates it was an existing group.
+     */
+    public CompletableFuture<Boolean> createGroup(String namespace, String group, GroupProperties groupProperties) {
+        Preconditions.checkArgument(group != null);
+        Preconditions.checkArgument(groupProperties != null);
+        Preconditions.checkArgument(validateRules(groupProperties.getSerializationFormat(), groupProperties.getCompatibility()));
+        log.info("create group called for {} with group properties {}", group, groupProperties);
+        return store.createGroup(namespace, group, groupProperties)
+                    .whenComplete((r, e) -> {
+                        if (e == null) {
+                            if (r) {
+                                log.info("Group {} created successfully.", group);
+                            } else {
+                                log.info("Group {} exists.", group);
+                            }
+                        } else {
+                            log.warn("create group {} request failed with error", group, e);
+                        }
+                    });
+    }
+
+    /**
+     * Gets group's properties.
+     * {@link GroupProperties#serializationFormat} which identifies the serialization format used to describe the schema.
+     * {@link GroupProperties#compatibility} sets the schema compatibility policy that needs to be enforced for evolving schemas.
+     * {@link GroupProperties#allowMultipleTypes} that specifies multiple schemas with distinct {@link SchemaInfo#type} can
+     * be registered.
+     * {@link GroupProperties#properties} properties for a group.
+     *
+     * @param namespace namespace for which the request is scoped to.
+     * @param group     Name of group.
+     * @return CompletableFuture which holds group properties upon completion.
+     */
+    public CompletableFuture<GroupProperties> getGroupProperties(String namespace, String group) {
+        Preconditions.checkArgument(group != null);
+        log.info("getGroupProperties called for group {}.", group);
+        return store.getGroupProperties(namespace, group)
+                    .whenComplete((r, e) -> {
+                        if (e == null) {
```

Comment:
Very similar code as that in last method. Opportunity for reuse via `extract method`?

Reply:
I am not sure i follow what you are asking for.. 
Say we add a whenComplete common code.. it would look something like following:
```
    public <T> CompletableFuture<T> whenComplete(CompletableFuture<T> future, Consumer<Throwable> throwableConsumer, 
                                                       Consumer<T> resultConsumer) {
        return future.whenComplete((r, e) -> {
            if (e != null) {
                throwableConsumer.accept(e);
            } else {
                resultConsumer.accept(r);
            }
        });
    }
```

and using it in the code will look as follows:
```
return whenCompleteLogger(store.getGroupProperties(namespace, group),
                e -> log.warn("getGroupProperties for group {} request failed with error", group, e),
                x -> log.debug("Group {} properties.", group));
```

This isnt very different from the original.

Reply:
```
return store.getGroupProperties(namespace, group)
                    .whenComplete((r, e) -> {
                        if (e == null) {
                            log.debug("Group {} {} properties.", namespace, group);
                        } else {
                            log.warn("getGroupProperties for group {} {} request failed with error", namespace, group, e);
                        }
                    });
```

It could be simplified though. For example: 

To:
```
log.debug("getGroupProperties called for group {} {}.", namespace, group);
return store.getGroupProperties(namespace, group)
                    .whenComplete((r, e) -> {
                        if (e != null) {
                            log.warn("getGroupProperties for group {} {} request failed with error", namespace, group, e);
                        }
                    });
```
In this case, if `getGroupProperties for group {} {} request failed with error` is missing, it'd mean the request was successful. If `getGroupProperties called for group {} {}.` is logged too, one can definitively tell the request was called too. So, there is a redundant condition and log statement. 

If that makes sense, there are other places where this pattern repeats, which can be simplified too.

Reply:
The debug level log tells us the processing was not just called but completed too. it helps to trace the flow to know which intermediate step we are at an what happened to it.